docs/proposals/auditing.md

```diff
@@ -0,0 +1,260 @@
+
+1. As a cluster operator I must be able to create a trace for **all accesses to a secret**.
+
+1. As a cluster operator I want to **define log-rotation** of a file-based output backend.
+
+1. As a cluster operator I must be able to create a trace for **kubectl exec** and other non-CRUD access to a container.
+
+1. As a cluster operator I must be able to get a trace of commands executed in **kubectl exec**. (TODO: out of scope?)
+
+## Community Work
+
+- Kubernetes basic audit log PR: https://github.com/kubernetes/kubernetes/pull/27087/ 
+- OpenStack's implementation of the CADF standard: https://www.dmtf.org/sites/default/files/standards/documents/DSP2038_1.1.0.pdf
+- Cloud Auditing Data Federation standard: https://www.dmtf.org/standards/cadf 
+- Ceilometer audit blueprint: https://wiki.openstack.org/wiki/Ceilometer/blueprints/support-standard-audit-formats 
+- Talk from IBM: An Introduction to DMTF Cloud Auditing using
+the CADF Event Model and Taxonomies https://wiki.openstack.org/w/images/e/e1/Introduction_to_Cloud_Auditing_using_CADF_Event_Model_and_Taxonomy_2013-10-22.pdf
+
+## Architecture
+
+When implementing audit logging there are basically two options:
+
+1. put a logging proxy in front of the apiserver
+2. integrate audit logging into the apiserver itself
+
+Both approaches have advantages and disadvanteges:
+
+- **pro proxy**:
+  + keeps complexity out of the apiserver
+  + reuses existing solutions
+- **contra proxy**:
+  + has no deeper insight into the Kubernetes api
+  + has no knowledge of auth, authn, admission
+  + has no access to the storage level for diffential output
+  + has to terminate SSL and complicates client certificates based auth
+
+In the following the second approach is described without a proxy.
+
+## Proposed Design
+
+The main concepts are those of
+- an audit *event*,
+- an audit *policy*,
+- an audit *policy action*
+- an audit *output backend*.
+
+An audit event holds all the data necessary for an *output backend* to produce a audit log entry. The *event* is independent of the *output backend*.
+
+The audit event struct is passed through the apiserver layers as an `*audit.Event` pointer inside the http context. It might be `nil` in case auditing is completely disabled.
+
+If auditing is enabled and the policy has an audit policy action (see below), the http handler will attach an `audit.Event` to the context:
+
+```go
+package api
+
+func WithAuditEvent(parent Context, e *audit.Event) Context
+func AuditEventFrom(ctx Context) (*audit.Event, bool)
+```
+
+Depending on the audit policy, different layers of the apiserver (e.g. http handler, storage) will fill the `audit.Event` struct. Certain fields might stay empty or `nil` if the policy does not require that field. E.g. in the case only http headers are supposed to be audit logged, no `OldObject` or `NewObject` is to be retrieved on the storage layer.
```

Comment:
I'd like to see a clear mapping of the audit policy to particular layers that will take part (will need modification) to provide necessary data.